pinot-spi/src/main/java/org/apache/pinot/spi/data/function/evaluators/ExpressionEvaluatorFactory.java

```diff
@@ -0,0 +1,111 @@
+   * 3. For columns ending with __KEYS or __VALUES (used for interpreting Map column in Avro), create default functions for handing the Map
+   * 4. Return null, if none of the above
+   */
+  @Nullable
+  public static ExpressionEvaluator getExpressionEvaluator(FieldSpec fieldSpec) {
+    ExpressionEvaluator expressionEvaluator = null;
+
+    String columnName = fieldSpec.getName();
+    String transformExpression = fieldSpec.getTransformFunction();
+    if (transformExpression != null) {
+
+      // if transform function expression present, use it to generate function evaluator
+      try {
+        expressionEvaluator = getExpressionEvaluator(transformExpression);
+      } catch (Exception e) {
+        LOGGER.error(
+            "Caught exception while constructing expression evaluator for transform expression: {}, of column: {}, skipping",
+            transformExpression, columnName, e);
+      }
+    } else if (fieldSpec.getFieldType().equals(FieldSpec.FieldType.TIME)) {
+
+      // for backward compatible handling of TIME field conversion
+      TimeFieldSpec timeFieldSpec = (TimeFieldSpec) fieldSpec;
+      TimeGranularitySpec incomingGranularitySpec = timeFieldSpec.getIncomingGranularitySpec();
+      TimeGranularitySpec outgoingGranularitySpec = timeFieldSpec.getOutgoingGranularitySpec();
+      if (outgoingGranularitySpec != null && !incomingGranularitySpec.equals(outgoingGranularitySpec)
+          && !incomingGranularitySpec.getName().equals(outgoingGranularitySpec.getName())) {
+        expressionEvaluator = new DefaultTimeSpecEvaluator(incomingGranularitySpec, outgoingGranularitySpec);
+      }
+    } else if (columnName.endsWith(SchemaFieldExtractorUtils.MAP_KEY_COLUMN_SUFFIX)) {
+
+      // for backward compatible handling of Map type (currently only in Avro)
+      String sourceMapName =
+          columnName.substring(0, columnName.length() - SchemaFieldExtractorUtils.MAP_KEY_COLUMN_SUFFIX.length());
+      String defaultMapKeysTransformExpression = getDefaultMapKeysTransformExpression(sourceMapName);
+      expressionEvaluator = getExpressionEvaluator(defaultMapKeysTransformExpression);
+    } else if (columnName.endsWith(SchemaFieldExtractorUtils.MAP_VALUE_COLUMN_SUFFIX)) {
+
+      // for backward compatible handling of Map type in avro (currently only in Avro)
+      String sourceMapName =
+          columnName.substring(0, columnName.length() - SchemaFieldExtractorUtils.MAP_VALUE_COLUMN_SUFFIX.length());
+      String defaultMapValuesTransformExpression = getDefaultMapValuesTransformExpression(sourceMapName);
+      expressionEvaluator = getExpressionEvaluator(defaultMapValuesTransformExpression);
+    }
+
+    return expressionEvaluator;
+  }
+
+  private static ExpressionEvaluator getExpressionEvaluator(String transformExpression) {
+    ExpressionEvaluator expressionEvaluator = null;
+    if (transformExpression != null && !transformExpression.isEmpty()) {
```

Comment:
(nit) transformExpression can never be null. Suggest moving this check to the caller